### PR TITLE
Add JavaScript Exponentiation

### PR DESCRIPTION
**Pull request by juancarlospaco**
- https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/Arithmetic_Operators#Exponentiation

### CODE REVIEW

**Comment by Yardanico** [comment]
Does this have any difference from `pow` in math library? Or maybe we can add this as `^` to math as well.

**Comment by juancarlospaco** [comment]
```nim
from math import pow
discard pow(2, 4)
```
 ⬆️ Nim ⬆️ &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ⬇️ Traceback ⬇️
```nim
temp.nim(1, 9) Error: type mismatch: got <int>
but expected one of:
proc pow(x, y: float32): float32
  first type mismatch at position: 1
  required type for x: float32
proc pow(x, y: float64): float64
  first type mismatch at position: 1
  required type for x: float64
```

---

```nim
from math import `^`
discard 2 ^ -4
```
 ⬆️ Nim ⬆️ &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ⬇️ Traceback ⬇️
```
temp.nim(1, 9) Error: conversion from int literal(-4) to Natural is invalid
```

```nim
discard 2 ** -4
```
 ⬆️ Nim ⬆️ &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ⬇️ JavaScript ⬇️
```javascript
/* Generated by the Nim Compiler v1.0.0 */
/*   (c) 2019 Andreas Rumpf */
var framePtr = null;
var excHandler = 0;
var lastJSError = null;
if (typeof Int8Array === 'undefined') Int8Array = Array;
if (typeof Int16Array === 'undefined') Int16Array = Array;
if (typeof Int32Array === 'undefined') Int32Array = Array;
if (typeof Uint8Array === 'undefined') Uint8Array = Array;
if (typeof Uint16Array === 'undefined') Uint16Array = Array;
if (typeof Uint32Array === 'undefined') Uint32Array = Array;
if (typeof Float32Array === 'undefined') Float32Array = Array;
if (typeof Float64Array === 'undefined') Float64Array = Array;
var nimvm_8110 = false;
var nim_program_result = 0;
var global_raise_hook_18618 = [null];
var local_raise_hook_18623 = [null];
var out_of_mem_hook_18626 = [null];
  if (!Math.trunc) {
    Math.trunc = function(v) {
      v = +v;
      if (!isFinite(v)) return v;
      return (v - v % 1)   ||   (v < 0 ? -0 : v === 0 ? v : 0);
    };
  }

(2 ** -4);
```

---

```nim
from math import `^`
discard 2 ^ 4
```
 ⬆️ Nim ⬆️ &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ⬇️ JavaScript ⬇️
```javascript
/* Generated by the Nim Compiler v1.0.0 */
/*   (c) 2019 Andreas Rumpf */
var framePtr = null;
var excHandler = 0;
var lastJSError = null;
if (typeof Int8Array === 'undefined') Int8Array = Array;
if (typeof Int16Array === 'undefined') Int16Array = Array;
if (typeof Int32Array === 'undefined') Int32Array = Array;
if (typeof Uint8Array === 'undefined') Uint8Array = Array;
if (typeof Uint16Array === 'undefined') Uint16Array = Array;
if (typeof Uint32Array === 'undefined') Uint32Array = Array;
if (typeof Float32Array === 'undefined') Float32Array = Array;
if (typeof Float64Array === 'undefined') Float64Array = Array;
var NTI57014 = {size: 0,kind: 31,base: null,node: null,finalizer: null};
var NTI57021 = {size: 0, kind: 18, base: null, node: null, finalizer: null};
var NTI3694 = {size: 0,kind: 31,base: null,node: null,finalizer: null};
var NTI3604 = {size: 0,kind: 20,base: null,node: null,finalizer: null};
NTI3604.base = NTI3694;
var NNI57021 = {kind: 2, len: 2, offset: 0, typ: null, name: null, sons: [{kind: 1, offset: "Field0", len: 0, typ: NTI57014, name: "Field0", sons: null}, 
{kind: 1, offset: "Field1", len: 0, typ: NTI3604, name: "Field1", sons: null}]};
NTI57021.node = NNI57021;

function setConstr() {
		    var result = {};
    for (var i = 0; i < arguments.length; ++i) {
      var x = arguments[i];
      if (typeof(x) == "object") {
        for (var j = x[0]; j <= x[1]; ++j) {
          result[j] = true;
        }
      } else {
        result[x] = true;
      }
    }
    return result;
  

	
}
var ConstSet1 = setConstr(17, 16, 4, 18, 27, 19, 23, 22, 21);

function nimCopy(dest_24827, src_24828, ti_24829) {
	var result_25019 = null;

		switch (ti_24829.kind) {
		case 21:
		case 22:
		case 23:
		case 5:
			if (!(is_fat_pointer_24801(ti_24829))) {
			result_25019 = src_24828;
			}
			else {
				result_25019 = [src_24828[0], src_24828[1]];
			}
			
			break;
		case 19:
			      if (dest_24827 === null || dest_24827 === undefined) {
        dest_24827 = {};
      }
      else {
        for (var key in dest_24827) { delete dest_24827[key]; }
      }
      for (var key in src_24828) { dest_24827[key] = src_24828[key]; }
      result_25019 = dest_24827;
    
			break;
		case 18:
		case 17:
			if (!((ti_24829.base == null))) {
			result_25019 = nimCopy(dest_24827, src_24828, ti_24829.base);
			}
			else {
			if ((ti_24829.kind == 17)) {
			result_25019 = (dest_24827 === null || dest_24827 === undefined) ? {m_type: ti_24829} : dest_24827;
			}
			else {
				result_25019 = (dest_24827 === null || dest_24827 === undefined) ? {} : dest_24827;
			}
			}
			nimCopyAux(result_25019, src_24828, ti_24829.node);
			break;
		case 24:
		case 4:
		case 27:
		case 16:
			      if (src_24828 === null) {
        result_25019 = null;
      }
      else {
        if (dest_24827 === null || dest_24827 === undefined) {
          dest_24827 = new Array(src_24828.length);
        }
        else {
          dest_24827.length = src_24828.length;
        }
        result_25019 = dest_24827;
        for (var i = 0; i < src_24828.length; ++i) {
          result_25019[i] = nimCopy(result_25019[i], src_24828[i], ti_24829.base);
        }
      }
    
			break;
		case 28:
			      if (src_24828 !== null) {
        result_25019 = src_24828.slice(0);
      }
    
			break;
		default: 
			result_25019 = src_24828;
			break;
		}

	return result_25019;

}
var nimvm_8110 = false;
var nim_program_result = 0;
var global_raise_hook_18618 = [null];
var local_raise_hook_18623 = [null];
var out_of_mem_hook_18626 = [null];
  if (!Math.trunc) {
    Math.trunc = function(v) {
      v = +v;
      if (!isFinite(v)) return v;

      return (v - v % 1)   ||   (v < 0 ? -0 : v === 0 ? v : 0);
    };
  }

function is_fat_pointer_24801(ti_24803) {
	var result_24804 = false;

	BeforeRet: do {
		result_24804 = !((ConstSet1[ti_24803.base.kind] != undefined));
		break BeforeRet;
	} while (false);

	return result_24804;

}

function nimCopyAux(dest_24832, src_24833, n_24835) {
		switch (n_24835.kind) {
		case 0:
			break;
		case 1:
			      dest_24832[n_24835.offset] = nimCopy(dest_24832[n_24835.offset], src_24833[n_24835.offset], n_24835.typ);
    
			break;
		case 2:
			    for (var i = 0; i < n_24835.sons.length; i++) {
      nimCopyAux(dest_24832, src_24833, n_24835.sons[i]);
    }
    
			break;
		case 3:
			      dest_24832[n_24835.offset] = nimCopy(dest_24832[n_24835.offset], src_24833[n_24835.offset], n_24835.typ);
      for (var i = 0; i < n_24835.sons.length; ++i) {
        nimCopyAux(dest_24832, src_24833, n_24835.sons[i][1]);
      }
    
			break;
		}

	
}

function HEX2AHEX3D_57024(x_57028, x_57028_Idx, y_57029) {
		x_57028[x_57028_Idx] = (x_57028[x_57028_Idx] * y_57029);

	
}

function HEX5E_57015(x_57018, y_57019) {
	var result_57020 = [0];

		switch (y_57019) {
		case 0:
			result_57020[0] = 1;
			break;
		case 1:
			result_57020[0] = x_57018;
			break;
		case 2:
			result_57020[0] = (x_57018 * x_57018);
			break;
		case 3:
			result_57020[0] = ((x_57018 * x_57018) * x_57018);
			break;
		default: 
			var colontmp__57116 = {Field0: 0, Field1: 0};
			var x_57022 = [0];
			var y_57023 = 0;
			nimCopy(colontmp__57116, {Field0: x_57018, Field1: nimCopy(null, y_57019, NTI3604)}, NTI57021);
			x_57022[0] = colontmp__57116["Field0"];
			y_57023 = colontmp__57116["Field1"];
			result_57020[0] = 1;
			L1: do {
					L2: while (true) {
					if (!true) break L2;
						if (!(((y_57023 & 1) == 0))) {
						HEX2AHEX3D_57024(result_57020, 0, x_57022[0]);
						}
						
						y_57023 = (y_57023 >> 1);
						if ((y_57023 == 0)) {
						break L1;
						}
						
						HEX2AHEX3D_57024(x_57022, 0, x_57022[0]);
					}
			} while(false);
			break;
		}

	return result_57020[0];

}


HEX5E_57015(2, 4);

```

```nim
discard 2 ** 4
```
 ⬆️ Nim ⬆️ &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ⬇️ JavaScript ⬇️
```javascript
/* Generated by the Nim Compiler v1.0.0 */
/*   (c) 2019 Andreas Rumpf */
var framePtr = null;
var excHandler = 0;
var lastJSError = null;
if (typeof Int8Array === 'undefined') Int8Array = Array;
if (typeof Int16Array === 'undefined') Int16Array = Array;
if (typeof Int32Array === 'undefined') Int32Array = Array;
if (typeof Uint8Array === 'undefined') Uint8Array = Array;
if (typeof Uint16Array === 'undefined') Uint16Array = Array;
if (typeof Uint32Array === 'undefined') Uint32Array = Array;
if (typeof Float32Array === 'undefined') Float32Array = Array;
if (typeof Float64Array === 'undefined') Float64Array = Array;
var nimvm_8110 = false;
var nim_program_result = 0;
var global_raise_hook_18618 = [null];
var local_raise_hook_18623 = [null];
var out_of_mem_hook_18626 = [null];
  if (!Math.trunc) {
    Math.trunc = function(v) {
      v = +v;
      if (!isFinite(v)) return v;
      return (v - v % 1)   ||   (v < 0 ? -0 : v === 0 ? v : 0);
    };
  }

(2 ** 4);
```

---

```nim
from math import pow
discard pow(2.0, 4.0)
```
 ⬆️ Nim ⬆️ &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ⬇️ JavaScript ⬇️
```javascript
/* Generated by the Nim Compiler v1.0.0 */
/*   (c) 2019 Andreas Rumpf */
var framePtr = null;
var excHandler = 0;
var lastJSError = null;
if (typeof Int8Array === 'undefined') Int8Array = Array;
if (typeof Int16Array === 'undefined') Int16Array = Array;
if (typeof Int32Array === 'undefined') Int32Array = Array;
if (typeof Uint8Array === 'undefined') Uint8Array = Array;
if (typeof Uint16Array === 'undefined') Uint16Array = Array;
if (typeof Uint32Array === 'undefined') Uint32Array = Array;
if (typeof Float32Array === 'undefined') Float32Array = Array;
if (typeof Float64Array === 'undefined') Float64Array = Array;
var nimvm_8110 = false;
var nim_program_result = 0;
var global_raise_hook_18618 = [null];
var local_raise_hook_18623 = [null];
var out_of_mem_hook_18626 = [null];
  if (!Math.trunc) {
    Math.trunc = function(v) {
      v = +v;
      if (!isFinite(v)) return v;
      return (v - v % 1)   ||   (v < 0 ? -0 : v === 0 ? v : 0);
    };
  }

Math.pow(2.0000000000000000e+00, 4.0000000000000000e+00);

```

---

Try for yourself using `nim js -f --nimcache:. -d:release -d:danger file.nim`
:slightly_smiling_face:

**Comment by Yardanico** [comment]
@juancarlospaco so yeah, why would you add it to a jsffi instead of adding it to the math  library?

**Comment by juancarlospaco** [comment]
The [link has a compatibility grid](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Arithmetic_Operators#Browser_compatibility).

The rest of the basic math operators are there, just this one was missing.

Math on JavaScript is a different Namespace, that is *not* on `math` neither, its on `jscore.nim` file.

`**` does not belong into the Math JavaScript Namespace, then is on `jsffi` with all other math operators without namespace.

Its how Nim itself is organized and its pre-existing to this PR.
:slightly_smiling_face:

File: lib/js/jsffi.nim [hunk 1]
```diff
@@ -184,6 +184,7 @@ proc `and`*(x, y: JsObject): JsObject {.importcpp: "(# && #)".}
 proc `or` *(x, y: JsObject): JsObject {.importcpp: "(# || #)".}
 proc `not`*(x:    JsObject): JsObject {.importcpp: "(!#)".}
 proc `in` *(x, y: JsObject): JsObject {.importcpp: "(# in #)".}
+proc `**` *(x, y: JsObject): JsObject {.importcpp: "(# ** #)".} ## https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/Arithmetic_Operators#Exponentiation
```
Comment (krux02) [inline lib/js/jsffi.nim]: The link makes the line too long. Put it on a different line or remove it entirely, I am not sure if we need it.

**Comment by krux02** [comment]
A test case is still missing.

**Comment by juancarlospaco** [comment]
Failing test is important_packages chronos.

**Comment by narimiran** [comment]
> Failing test is important_packages chronos.

Yep, unrelated CI failure.

File: tests/js/tjsffi.nim [hunk 2]
```diff
@@ -331,3 +349,25 @@ block:
...
+  echo 32767 ** 2 == 1073676289
+  echo 0.0 ** 0.0 == 1
+  echo 1.0 ** 1.0 == 1
+  echo -1.0 ** -1.0 == -1
+  echo 9.0 ** 4.0 == 6561
+  echo 4.0 ** 9.0 == 262144
+  echo -9.0 ** 9.0 == -387420489
+  echo 42.0 ** 2.0 == 1764
+  echo 127.0 ** 2.0 == 16129
+  echo 32767.0 ** 2.0 == 1073676289
```
Comment (krux02) [inline tests/js/tjsffi.nim]: You are not operating on `JsObject` anymore. This is not what I originally accepted.

File: lib/js/jsffi.nim [hunk 3]
```diff
@@ -184,6 +184,7 @@ proc `and`*(x, y: JsObject): JsObject {.importcpp: "(# && #)".}
 proc `or` *(x, y: JsObject): JsObject {.importcpp: "(# || #)".}
 proc `not`*(x:    JsObject): JsObject {.importcpp: "(!#)".}
 proc `in` *(x, y: JsObject): JsObject {.importcpp: "(# in #)".}
+proc `**` *(x, y: SomeNumber): SomeNumber {.importcpp: "((#) ** #)".}
```
Comment (krux02) [inline lib/js/jsffi.nim]: What? Originally you did this on `JsObject`, now it is on ``SomeNumber``. If you just want to provide the nim mapping of what is a builtin in javascript, you have to use JsObject everywhere, here as well.

If you want to provide exponentiation, we already have ``math.pow``.

Comment (andreaferretti) [reply lib/js/jsffi.nim]: I don't understand this feedback. Exponentiation does not work on `JsObject`. In javascript you do not exponentiate two objects, you exponentiate two numbers.

**Pull request closed without merging**